### PR TITLE
feat: merge rule.meta.defaultOptions before validation

### LINKED ISSUE #165: Fix: Validate options after applying defaults, not before
Coming over from https://github.com/eslint/eslint/pull/17656#issuecomment-2027532264: right now, eslintrc validates options and _then_ validates defaults. Copying the example from @mdjermanovic's comment, if this is a rule:

```js
{
    meta: {
        defaultOptions: [{
            foo: 42
        }],
        schema: [{
            type: "object",
            maxProperties: 2 // allows one or two in addition to "foo"?
        }]
    },
    create() {
        return {};
    }
}
```

...then the configuration `["error", { bar: 6, baz: 7 }]` is still considered valid.

This is different from the new ESLint flat config format.

Technically this proposed stricter behavior could be considered a breaking change and out of scope given this package's feature freeze. However, I'd propose we consider it a bugfix, as IMO the invalid options should not have been allowed to begin with.

Filing a tracking issue for my reference. I plan on sending a PR for this soon, unless directed otherwise.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Fixes #165.

I felt unsure labeling this as "feat" vs. "fix" - the root behavior is, arguably, a _bug fix_ for _feature support_ of `rule.meta.defaultOptions`... 🤷 

Code roughly copied from https://github.com/eslint/eslint/pull/17656.

### CODE REVIEW

File: lib/shared/config-validator.js
Comment (nzakas) [inline lib/shared/config-validator.js:152]: We can't use `?` in this package because it's also used by ESLint v8, which supports older Node.js versions that don't have support for this syntax.

**Review by nzakas — DISMISSED** [review]
Overall LGTM. Just one note about syntax.

Would like @mdjermanovic to review before merging.

**Review by nzakas — DISMISSED** [review]

**Comment by nzakas** [comment]
@eslint/eslint-tsc This PR is still looking for another review.

File: tests/lib/shared/deep-merge-arrays.js
Comment (fasttime) [inline tests/lib/shared/deep-merge-arrays.js:1]: I believe the intent of `no-undefined` is to let you write `void 0` instead of `undefined`. It won't restrict you from using `null`.

**Review by fasttime — COMMENTED** [review]
Sorry for the late review. I just left a few suggestions. Other than that, I think this looks good.

File: tests/lib/shared/deep-merge-arrays.js
Comment (JoshuaKGoldberg) [reply tests/lib/shared/deep-merge-arrays.js:1]: `void 0` is ... not a practice I'm used to anybody preferring 😄. I'm unclear on what the suggested next step is.

Comment (fasttime) [reply tests/lib/shared/deep-merge-arrays.js:1]: We use `void 0` instead of `undefined` in the rest of the repo, so we could do the same thing in this file.

File: lib/shared/config-validator.js
Comment (fasttime) [reply lib/shared/config-validator.js:152]: The minimum Node.js version currently supported is [18.18.0](https://github.com/eslint/eslintrc/blob/v3.1.0/package.json#L80), which does recognize the optional chain operator syntax `?.`.

File: tests/lib/shared/deep-merge-arrays.js
Comment (ljharb) [reply tests/lib/shared/deep-merge-arrays.js:1]: it's typically/historically preferred to ack that `undefined` can be redefined, and `void 0` is shorter than `void undefined` (altho i've taken to using the latter since it's more readable and probably equally minifiable)

**Review by fasttime — APPROVED** [review]
LGTM, thanks! Pending a last review from @mdjermanovic.

**Review by mdjermanovic — COMMENTED** [review]
This LGTM, but I think we should hold off merging until https://github.com/eslint/eslint/pull/17656 is approved to make sure the changes are synchronized.

**Pull request merged**
